modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/workflow/processors/SW360EnricherImpl.java

```diff
@@ -73,11 +75,14 @@ private void mapExternalIdsOnSW360Release(SW360Release sw360Release, Artifact ar
         mapCoordinates(sw360Release)
                 .forEach(artifact::addFact);
 
-        artifact.addFact(new DeclaredLicenseInformation(makeLicenseStatementFromString(sw360Release.getDeclaredLicense())));
-        artifact.addFact(new ObservedLicenseInformation(makeLicenseStatementFromString(sw360Release.getObservedLicense())));
+        addLicenseFact(Optional.ofNullable(sw360Release.getDeclaredLicense()), artifact, DeclaredLicenseInformation::new);
+        addLicenseFact(Optional.ofNullable(sw360Release.getObservedLicense()), artifact, ObservedLicenseInformation::new);
         artifact.addFact(new ArtifactReleaseTagURL(sw360Release.getReleaseTagUrl()));
```

Comment:
should the other data variables not also be put into an optional? (e.g. CopyrightStatement and ArtifactReleaseTagURL)

Reply:
Good point

Reply:
facts are build in a way, that every fact implements an `.isEmpty()` method. And if it is empty it is treated similar to a `Optional.empty()` and adding an empty fact is a no-op.

Reply:
Does this apply to all of them or is there another one with special creation conditions? Since there have been issues with empty softwareheritage id and license? Even though they are special cases because they have special creation conditions.

Reply:
All facts implement that method:

https://github.com/eclipse/antenna/blob/a89d2bddace67078617296b2084ae8bfcb70bc81/core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/ArtifactFact.java#L16

Reply:
From what I saw, adding an empty fact is not a noop, because if I do not add it, it is not there and I do not have a problem, with empty facts I have to handle explicitely that the fact is empty and this does not make sense, because if I ask for a fact, I would expect that the fact is either there and has some content or is not returned at all.

Reply:
SO if I look at the code, it would make sense to not add the fact in org.eclipse.sw360.antenna.model.artifact.ArtifactCore#addFact(org.eclipse.sw360.antenna.model.artifact.ArtifactFact), i.e., if fact.isEmpty is true. Then I would agree that it is a noop and I would think that the askFor methods would work as expected.

Reply:
Yes. It is the other way round. One can add an empty fact and thus explicitly say that this statement should be emtpy. So adding is not a no-op. Only from the view of an empty artifact. Adding an empty fact is equal to "there is no fact at all". But if there would be a previous fact it would get overwritten.

Reply:
Ok, but then, it makes sense to not add facts in the sw360 enricher because the semantics is not that if sw360 does not know a fact, it is not there, only that sw360 does not know whether there is a fact.